### ISSUE TITLE
Export downloaded files

### ISSUE THREAD

**Issue by totchi-lagawi** [OuterTune/OuterTune#106]
### Checklist

- [X] I've checked that there is no other issue about this feature request.
- [X] This issue contains only one feature request.
- [X] The title of this issue accurately describes the feature request.

### Feature description

This was asked for both the original InnerTube and its fork by Malopieds, and since this fork seems to be much more powerful than those, I ask it here too.

As the title supposed, I'd like to be able to export the downloaded songs, in for example `/storage/emulated/UID/Music`, and of course in the MPEG-3 or FLAC format. As of the local storage support of OuterTune, it could even be possible to play these songs from within the app. It could even be considered to replace the current download system with this one.

Whatever you choose to do, thanks for this app. InnerTune is the best music playing app I found so far, and the lack of some functionalities was the only problem I could find about it. I have some, well, troubles with Java and Gradle/Maven/Android Studio, and so I am thankful to you for having brought these in this app. Thanks.

### Why do you want this feature?

I have a MP3 player, which can as of its nature only read audio files. Manually downloading files from YouTube works, but it is a pain, as well as moving the files from my computer to my SD card. Be that as it may, having songs stored externally has numerous advantages, like keeping songs even after an app reinstall or just a phone reset.

### Additional information

Thanks again!

**Comment by Ostrym** [comment]
This would be great !

**Comment by aasutossh** [comment]
So right now it simply caches the song? We can't download the "file" into internal storage?

**Comment by totchi-lagawi** [comment]
Right now, when we download a song, it is actually downloaded, but in 
the internal storage of the application, which isn't accessible without 
rooting the device (and goodbye warranty...). Furthermore, it isn't 
downloaded in a conventional format like MPEG-3 but in a Google-specific 
format (EXO)…

**Comment by totchi-lagawi** [comment]
So even if these files are recovered they still need to be converted to 
be playable almost everywhere (and metadata should also be added to 
them, since the application keeps them in a separate database). At 
least, downloaded music doesn't seem to be split between multiple files 
and encrypted since VLC can play the files without any problem.

**Comment by mikooomich** [comment]
Yes. .exo is essentially opus codec in a webm container (ofc, this can vary depending on the file YT serves), and there's no drm or anything (well, for now)

I could add something that just exports downloads to internal storage... but it seems stupid to copy files twice, when we already have local media support anyways. Wanted to create a system to download files to /sdcard, but haven't gotten around to that yet

**Comment by totchi-lagawi** [comment]
Modifying the current download system to download to `/sdcard` is the 
best way to implement this, I think.

**Comment by totchi-lagawi** [comment]
|/sdcard, I didn't know GitHub doesn't support Markdown over e-mails...
|

**Comment by totchi-lagawi** [comment]
GitHub, please improve e-mail reply.

**Comment by mikooomich** [comment]
> Modifying the current download system to download to `/sdcard` is the 
> best way to implement this, I think.
> 
> 

You can't. Simple cache (what we use rn) is only able to write in the apps internal storage, not /sdcard.

I played around with a custom download implementation, and while it does download and play offline, that's about it. Alot of ui stuff and nuances still need to be worked out (if anyone wants to help out, please ping me).

I don't think newer Android versions let you write media files just anywhere in /sdcard though, so keep on mind that limitation. It needs to be in certain folders like /Music

**Comment by mikooomich** [comment]
#

**Comment by romeo1250-crypto** [comment]
I am planning to fork the repo and build a UI filetype selector with a MediaStore export pipeline to fix this. The UI selector, will b elike that of snaptube, where the user can select the filetype he wants, mp3, opus, and other audio filetypes where applicable...
 Is anyone currently working on it?
